diff --git a/Generics.agda b/Generics.agda
index 4586617..c5795fe 100644
--- a/Generics.agda
+++ b/Generics.agda
@@ -8,14 +8,14 @@ open import Level hiding (suc)
 open import Function
 
 open import Reflection hiding (visibility)
-open import Reflection.Term hiding (getName)
-import Reflection.Name as Name
-open import Reflection.Pattern
-import Reflection.Abstraction as Abs
-open import Reflection.Argument as Arg hiding (map)
-open import Reflection.Argument.Information
-open import Reflection.Argument.Visibility as Vis
-open import Reflection.TypeChecking.Monad.Syntax
+open import Reflection.AST.Term hiding (getName)
+import Reflection.AST.Name as Name
+open import Reflection.AST.Pattern
+import Reflection.AST.Abstraction as Abs
+open import Reflection.AST.Argument as Arg hiding (map)
+open import Reflection.AST.Argument.Information
+open import Reflection.AST.Argument.Visibility as Vis
+open import Reflection.TCM.Syntax
 
 open import Data.Unit
 open import Data.Nat
